core/change.go

```
@@ -100,79 +109,133 @@ func (change Change) Write(writer io.Writer) error {
+	}
 
-		_, change.Component, err = compPrompt.Run()
-		if err != nil {
-			return err
-		}
+	err := change.promptForComponent(&ctx)
+
+	if err != nil {
+		return err
 	}
 
-	if len(config.Kinds) > 0 {
-		kindPrompt := promptui.Select{
-			Label: "Kind",
-			Items: config.Kinds,
-			Stdin: stdinReader,
-		}
+	err = change.promptForKind(&ctx)
 
-		_, change.Kind, err = kindPrompt.Run()
-		if err != nil {
-			return err
-		}
+	if err != nil {
+		return err
+	}
 
-		kind = kindFromLabel(config, change.Kind)
+	err = change.parseKind(&ctx)
+
+	if err != nil {
+		return err
 	}
 
-	if kind == nil || !kind.SkipBody {
-		bodyPrompt := config.Body.CreatePrompt(stdinReader)
-		change.Body, err = bodyPrompt.Run()
+	err = change.promptForBody(&ctx)
 
-		if err != nil {
-			return err
+	if err != nil {
+		return err
+	}
+
+	return change.promptForUserChoices(&ctx)
+}
+
+func (change *Change) promptForComponent(ctx *PromptContext) error {
+	if len(ctx.config.Components) == 0 {
+		return nil
+	}
+
+	compPrompt := promptui.Select{
+		Label: "Component",
+		Items: ctx.config.Components,
+		Stdin: ctx.stdinReader,
+	}
+
+	var err error
+	_, change.Component, err = compPrompt.Run()
+
+	return err
+}
+
+func (change *Change) promptForKind(ctx *PromptContext) error {
+	if len(ctx.config.Kinds) == 0 || len(change.Kind) > 0 {
+		return nil
+	}
+
+	kindPrompt := promptui.Select{
+		Label: "Kind",
+		Items: ctx.config.Kinds,
+		Stdin: ctx.stdinReader,
+	}
+
+	var err error
+	_, change.Kind, err = kindPrompt.Run()
+
+	return err
+}
+
+func (change *Change) parseKind(ctx *PromptContext) error {
+	if len(change.Kind) == 0 {
+		return nil
+	}
+
+	for i := range ctx.config.Kinds {
+		kindConfig := &ctx.config.Kinds[i]
+
+		if kindConfig.Label == change.Kind {
+			ctx.kind = kindConfig
+			return nil
 		}
 	}
 
+	return ErrInvalidKind{change.Kind}
+}
+
+func (change *Change) promptForBody(ctx *PromptContext) error {
+	if (ctx.kind == nil || !ctx.kind.SkipBody) && len(change.Body) == 0 {
```

There is another edge case here where a body was supplied but the kind is not meant to have one, we should return an error for that.
`if ctx.kind != nil && ctx.kind.SkipBody && len(change.Body) > 0 { bad }`

---

Good point, I added a test and fixed it too.